File: docs/integration-examples/python/logger.py
"""
DevSmith Logger - Python Integration

Copy this file into your project and customize the configuration.

Usage:
    from logger import DevSmithLogger
    
    logger = DevSmithLogger(
        api_key=os.getenv('DEVSMITH_API_KEY'),
        api_url=os.getenv('DEVSMITH_API_URL', 'http://localhost:3000'),
        project_slug='my-project',
        service_name='api-server'
    )
    
    logger.info('User logged in', user_id=123)
    logger.error('Database error', code='ECONNREFUSED')
"""

import requests
import time
import atexit
import os
import threading
from datetime import datetime
from typing import Dict, Any, Optional

class DevSmithLogger:
    def __init__(
        self,
        api_key: str,
        api_url: str,
        project_slug: str,
        service_name: str,
        batch_size: int = 100,
        flush_interval: float = 5.0  # seconds
    ):
        # Validate required config
        if not api_key:
            raise ValueError('DevSmithLogger: api_key is required')
        if not project_slug:
            raise ValueError('DevSmithLogger: project_slug is required')
        if not service_name:
            raise ValueError('DevSmithLogger: service_name is required')
        
        self.api_key = api_key
        self.api_url = api_url.rstrip('/')
        self.project_slug = project_slug
        self.service_name = service_name
        self.batch_size = batch_size
        self.flush_interval = flush_interval
        
        self.buffer = []
        self.lock = threading.Lock()
        self.timer = None
        
        # Setup flush timer
        self._start_timer()
        
        # Flush on exit
        atexit.register(self.flush)
    
    def _start_timer(self):
        """Start periodic flush timer"""
        def flush_periodically():
            with self.lock:
                if self.buffer:
                    self.flush()
            self._start_timer()
        
        self.timer = threading.Timer(self.flush_interval, flush_periodically)
        self.timer.daemon = True
        self.timer.start()
    
    def log(self, level: str, message: str, **context):
        """Add log entry to buffer"""
        entry = {
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'level': level.upper(),
            'message': message,
            'service': self.service_name,
            'context': context
        }
        
        with self.lock:
            self.buffer.append(entry)
            
            # Flush if batch size reached
            if len(self.buffer) >= self.batch_size:
                self.flush()
    
    def flush(self):
        """Send buffered logs to DevSmith API"""
        with self.lock:
            if not self.buffer:
                return
            
            logs = self.buffer[:]
            self.buffer.clear()
        
        payload = {
            'project_slug': self.project_slug,
            'logs': logs
        }
        
        headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {self.api_key}'
        }
        
        try:
            response = requests.post(
                f'{self.api_url}/api/logs/batch',
                json=payload,
                headers=headers,
                timeout=10
            )
            
            if response.status_code not in (200, 201):
                print(f'DevSmith Logger: Failed to send logs ({response.status_code}): {response.text}')
                # Re-add logs to buffer for retry
                with self.lock:
                    self.buffer.extend(logs)
        
        except requests.exceptions.RequestException as e:
            print(f'DevSmith Logger: Network error: {e}')
            # Re-add logs to buffer for retry
            with self.lock:
                self.buffer.extend(logs)
    
    # Convenience methods
    def debug(self, message: str, **context):
        self.log('DEBUG', message, **context)
    
    def info(self, message: str, **context):
        self.log('INFO', message, **context)
    
    def warn(self, message: str, **context):
        self.log('WARN', message, **context)
    
    def error(self, message: str, **context):
        self.log('ERROR', message, **context)

File: docs/integration-examples/javascript/logger.js
/**
 * DevSmith Logger - JavaScript/Node.js Integration
 * 
 * Copy this file into your project and customize the configuration.
 * 
 * Usage:
 *   const DevSmithLogger = require('./logger');
 *   const logger = new DevSmithLogger({
 *     apiKey: process.env.DEVSMITH_API_KEY,
 *     apiUrl: process.env.DEVSMITH_API_URL || 'http://localhost:3000',
 *     projectSlug: 'my-project',
 *     serviceName: 'api-server'
 *   });
 *   
 *   logger.info('User logged in', { userId: 123 });
 *   logger.error('Database error', { code: 'ECONNREFUSED' });
 */

const https = require('https');
const http = require('http');

class DevSmithLogger {
  constructor(config) {
    this.apiKey = config.apiKey;
    this.apiUrl = config.apiUrl || 'http://localhost:3000';
    this.projectSlug = config.projectSlug;
    this.serviceName = config.serviceName;
    this.batchSize = config.batchSize || 100;
    this.flushInterval = config.flushInterval || 5000; // 5 seconds
    
    this.buffer = [];
    this.timer = null;
    
    // Validate required config
    if (!this.apiKey) {
      throw new Error('DevSmithLogger: apiKey is required');
    }
    if (!this.projectSlug) {
      throw new Error('DevSmithLogger: projectSlug is required');
    }
    if (!this.serviceName) {
      throw new Error('DevSmithLogger: serviceName is required');
    }
    
    // Setup flush timer
    this.startTimer();
    
    // Flush on process exit
    process.on('exit', () => this.flush());
    process.on('SIGINT', () => {
      this.flush();
      process.exit();
    });
  }
  
  startTimer() {
    this.timer = setInterval(() => {
      if (this.buffer.length > 0) {
        this.flush();
      }
    }, this.flushInterval);
  }
  
  log(level, message, context = {}) {
    const entry = {
      timestamp: new Date().toISOString(),
      level: level.toUpperCase(),
      message: message,
      service: this.serviceName,
      context: context
    };
    
    this.buffer.push(entry);
    
    // Flush if batch size reached
    if (this.buffer.length >= this.batchSize) {
      this.flush();
    }
  }
  
  flush() {
    if (this.buffer.length === 0) {
      return;
    }
    
    const logs = [...this.buffer];
    this.buffer = [];
    
    const payload = JSON.stringify({
      project_slug: this.projectSlug,
      logs: logs
    });
    
    const url = new URL('/api/logs/batch', this.apiUrl);
    const isHttps = url.protocol === 'https:';
    const client = isHttps ? https : http;
    
    const options = {
      hostname: url.hostname,
      port: url.port || (isHttps ? 443 : 80),
      path: url.pathname,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(payload),
        'Authorization': `Bearer ${this.apiKey}`
      }
    };
    
    const req = client.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => {
        data += chunk;
      });
      res.on('end', () => {
        if (res.statusCode !== 200 && res.statusCode !== 201) {
          console.error(`DevSmith Logger: Failed to send logs (${res.statusCode}):`, data);
        }
      });
    });
    
    req.on('error', (error) => {
      console.error('DevSmith Logger: Network error:', error.message);
      // Re-add logs to buffer for retry
      this.buffer.push(...logs);
    });
    
    req.write(payload);
    req.end();
  }
  
  // Convenience methods
  debug(message, context) {
    this.log('DEBUG', message, context);
  }
  
  info(message, context) {
    this.log('INFO', message, context);
  }
  
  warn(message, context) {
    this.log('WARN', message, context);
  }
  
  error(message, context) {
    this.log('ERROR', message, context);
  }
}

module.exports = DevSmithLogger;
